### PR TITLE
1786 improve user migrator

### PR DESCRIPTION
**Pull request by amartinj**
- Log less frequently
- DRY imports
Related to [#293](https://github.com/CartoDB/tech-ops/issues/293)
Related to [#1786](https://github.com/CartoDB/cartodb-central/issues/1786)

Migration: #13108

### CODE REVIEW

File: services/user-mover/utils.rb [hunk 1]
```diff
@@ -60,10 +60,12 @@ def remove_user_mover_banner(user_id)
       def run_command(cmd)
         logger.debug "Running command: \"#{cmd}\""
         return_code = nil
+        log_message = ''
         Open3.popen2e(cmd) do |_stdin, stdout_and_stderr, wait_thr|
-          stdout_and_stderr.each { |line| logger.debug line.strip }
+          stdout_and_stderr.each { |line| log_message = log_message + line.strip + "\n" }
```
Comment (javitonino) [inline services/user-mover/utils.rb]: Have you seen the `SAVE_BLOCK` constant at `Carto::Log::LogWriter`? In both cases, we are keeping the whole log in memory, but via manipulating the constant we get regular status updates by looking at the DB (at the cost of slowness). IMHO: we should have some way to check the in-progress log, it's very useful.

An alternative we mentioned could be to write to a temp file each line, and at the end of the import, copy the file into the database.

Let's discuss!

File: services/user-mover/import_user.rb [hunk 2]
```diff
@@ -579,8 +579,10 @@ def configure_database(target_dbhost)
         # Note: this will change database_host on the user model to perform configuration but will not actually store
         # the change
         user_model = ::User.find(username: @target_username)
-        user_model.database_host = target_dbhost
-        user_model.database_name = @target_dbname
+        unless @options[:dry]
+          user_model.database_host = target_dbhost
+          user_model.database_name = @target_dbname
+        end
```
Comment (javitonino) [inline services/user-mover/import_user.rb]: I think this is not enough. As the comment suggests, this change is not persisted to the database. However, we see the database being modified.

I think this is done here: https://github.com/CartoDB/cartodb/blob/master/services/user-mover/import_user.rb#L545

Also, as we already do the update from central at the end, I would try to remove this change from the import job always. That is, this job never updates the database host, and the switch is always done via central. That way, the `dry` import can be done by running the box task and not running the central task.

What do you think?

File: app/models/carto/user_migration_import.rb [hunk 3]
```diff
@@ -132,6 +132,7 @@ def rollback_import_data(package)
     end
 
     def update_database_host
+      return if dry
```
Comment (javitonino) [inline app/models/carto/user_migration_import.rb]: What about moving this conditional to the caller of this function? This way, we can manually run this function if needed, without touching the `dry` parameter.

File: services/user-mover/utils.rb [hunk 4]
```diff
@@ -60,10 +61,18 @@ def remove_user_mover_banner(user_id)
       def run_command(cmd)
         logger.debug "Running command: \"#{cmd}\""
         return_code = nil
+        log_message = ''
+        file = Tempfile.new('datamover')
```
Comment (javitonino) [inline services/user-mover/utils.rb]: I'd recommend using [`open`](https://ruby-doc.org/stdlib-2.2.3/libdoc/tempfile/rdoc/Tempfile.html#method-c-new-label-Exceptions) with a block, in order to clearly mark the tempfile lifecycle.

File: services/user-mover/import_user.rb [hunk 5]
```diff
@@ -530,6 +530,7 @@ def setup_db(_dbname)
       end
 
       def update_database_retries(userid, username, db_host, db_name, retries = 1)
+        return if @options[:dry]
```
Comment (javitonino) [inline services/user-mover/import_user.rb]: Same comment, maybe put the conditional in the caller.

File: db/migrate/20171113180900_add_dry_to_user_migration_import.rb [hunk 6]
```diff
@@ -0,0 +1,12 @@
+require 'carto/db/migration_helper'
+
+include Carto::Db::MigrationHelper
+
+migration(
+  Proc.new do
+    add_column :user_migration_imports, :dry, :boolean
```
Comment (javitonino) [inline db/migrate/20171113180900_add_dry_to_user_migration_import.rb]: Nullable? Let's make this NOT NULL with a default.

File: app/models/carto/user_migration_import.rb [hunk 7]
```diff
@@ -167,14 +168,15 @@ def import_job_arguments(data_dir)
...
+        import_job_logger: log.logger,
+        dry: dry
       }
     end
 
     def set_defaults
       self.log = Carto::Log.create(type: 'user_migration_import') unless log
       self.state = STATE_PENDING unless state
-
+      self.dry = true unless dry
```
Comment (javitonino) [inline app/models/carto/user_migration_import.rb]: This always sets `dry` to true. If it was set to false, the conditional fails, and it overwrites with `true`.

File: services/user-mover/import_user.rb [hunk 8]
```diff
@@ -17,7 +17,7 @@ class ImportJob
       attr_reader :logger
 
       def initialize(options)
-        default_options = { data: true, metadata: true, set_banner: true, update_metadata: true }
+        default_options = { data: true, metadata: true, set_banner: true, update_metadata: true, dry: false }
```
Comment (javitonino) [inline services/user-mover/import_user.rb]: Is this flag related to `update_metadata`? The name suggests it might do the same.

File: app/models/carto/user_migration_import.rb (hunk 7 above)
Comment (amartinj) [reply app/models/carto/user_migration_import.rb]: simply changed to true. my bad

File: services/user-mover/utils.rb (hunk 4 above)
Comment (amartinj) [reply services/user-mover/utils.rb]: Using a block would cause three nested blocks. Better closing it explicitly when the method ends

**Comment by amartinj** [comment]
Please do not merge. Let me know if the CR is fine so I can add some more tests. I was not sure of the approach so I found it useless to add tests beforehand.

**Comment by juanignaciosl** [comment]
Added "do not merge yet" label accordingly to @amartinj comment: "Please do not merge" :-) Beginning the CR...

File: app/models/carto/user_migration_import.rb [hunk 9]
```diff
@@ -83,7 +83,7 @@ def do_import_metadata(package, service)
         raise e
       end
       org_import? ? self.organization = imported : self.user = imported
-      update_database_host
+      update_database_host unless dry
```
Comment (juanignaciosl) [inline app/models/carto/user_migration_import.rb]: But, if you only use `dry` here, you'd import visualizations as well, isn't it? Check `import` method. I think that the check should be there, avoiding `do_import_metadata` and `import_visualizations`. `import_metadata?` might deserve a wrapper including `import_metadata? && dry`, because there's a third usage (`service.rollback_import_from_directory`) that might need it as well.

If you agree, please write a (failing) test about (not) importing visualizations before fixing it.

File: services/user-mover/import_user.rb [hunk 10]
```diff
@@ -200,7 +200,7 @@ def import_user
           end
         end
 
-        if @options[:data]
+        if @options[:data] && @options[:update_metadata]
```
Comment (juanignaciosl) [inline services/user-mover/import_user.rb]: Is this true? I think that `update_metadata` refers to "all" metadata (the full user and so on). Keep in mind that the database host must change on data change. If you're moving it outside (we spoke about that), is it implemented somewhere else?

File: services/user-mover/utils.rb (hunk 4 above)
Comment (juanignaciosl) [reply services/user-mover/utils.rb]: I also prefer a block, even if it requires extracting a new method for legibility. Minor, though.

**Comment by juanignaciosl** [comment]
BTW, this PR had to have "do not merge yet" label from the very beginning because it contains a migration.

File: app/models/carto/user_migration_import.rb (hunk 9 above)
Comment (amartinj) [reply app/models/carto/user_migration_import.rb]: The way I understood it is that ```dry``` was some sort of ```do not touch database references``` but nothing about not importing metadata :S

Comment (juanignaciosl) [reply app/models/carto/user_migration_import.rb]: For future readers, no. This doubt has been caused by an underspecification of what dry means at the issue, sorry, my bad. I just added this to [the key comment](https://github.com/CartoDB/tech-ops/issues/293#issuecomment-3435233249): "This dry migration means that the importer will only import the DB dump, and only that (no Redis, no metadata change, no anything else).".

**Comment by juanignaciosl** [comment]
Additional note: please update `user_migrator.html` documentation if mover behaviour changes. For example, it will probably have to explain how to use DRY feature or update database host manually at box and central if you use the mover.

**Comment by amartinj** [comment]
retest this, please

**Comment by amartinj** [comment]
Reteset this, please

**Comment by amartinj** [comment]
Retest this, please

File: services/user-mover/utils.rb [hunk 11]
```diff
@@ -59,11 +60,21 @@ def remove_user_mover_banner(user_id)
...
+      end
+
+      def run_command_with_log(cmd, file)
         return_code = nil
+        log_message = ''
         Open3.popen2e(cmd) do |_stdin, stdout_and_stderr, wait_thr|
-          stdout_and_stderr.each { |line| logger.debug line.strip }
+          stdout_and_stderr.each do |line|
+            message = line.strip + "\n"
+            log_message = log_message + message
```
Comment (juanignaciosl) [inline services/user-mover/utils.rb]: Minor: replace `log_message = log_message + message` with `log_message += message`.

File: services/user-mover/utils.rb [hunk 12]
```diff
@@ -59,11 +60,21 @@ def remove_user_mover_banner(user_id)
 
       def run_command(cmd)
         logger.debug "Running command: \"#{cmd}\""
+        Tempfile.open('datamover') { |f| run_command_with_log(cmd, f) }
```
Comment (juanignaciosl) [inline services/user-mover/utils.rb]: I don't get the usage of `Tempfile`:
- if the command works, the messages are dumped into the DB and the file is deleted. And messages are accumulated in memory anyway, so there's no memory improvement.
- if the command fails, the file is deleted once it's garbage collected, so it can't be read.

In both cases it's not useful. Am I missing something?

If it's needed for some reason, check [best practices about closing it](https://ruby-doc.org/stdlib-1.9.3/libdoc/tempfile/rdoc/Tempfile.html#class-Tempfile-label-Explicit+close).

Comment (javitonino) [reply services/user-mover/utils.rb]: The motivation is that I asked for some way to track progress status while the migration is running. I find it useful to look at `pg_restore` output to estimate remaining time.

File: db/migrate/20171113180900_add_dry_to_user_migration_import.rb [hunk 13]
```diff
@@ -0,0 +1,12 @@
+require 'carto/db/migration_helper'
+
+include Carto::Db::MigrationHelper
+
+migration(
+  Proc.new do
+    add_column :user_migration_imports, :dry, :boolean, default: true
```
[the remark this answers is not in the document]
Comment (amartinj) [reply db/migrate/20171113180900_add_dry_to_user_migration_import.rb]: I will invoke @javitonino for this. You are writing my spoken words 🤣

File: app/models/carto/user_migration_import.rb [hunk 14]
```diff
@@ -167,14 +168,14 @@ def import_job_arguments(data_dir)
         into_org_name: nil,
         mode: :import,
         logger: log.logger,
-        import_job_logger: log.logger
+        import_job_logger: log.logger,
+        import_metadata: !dry
```
[the remark this answers is not in the document]
Comment (amartinj) [reply app/models/carto/user_migration_import.rb]: The default for ```import_metadata``` is true at the (user mover)[https://github.com/CartoDB/cartodb/blob/master/services/user-mover/import_user.rb#L20]. It is doing _some_ metadata updating there. @javitonino and I spoke about removing that responsibility from there and leave it only with data restoring.

File: db/migrate/20171113180900_add_dry_to_user_migration_import.rb (hunk 13 above)
[the remark this answers is not in the document]
Comment (javitonino) [reply db/migrate/20171113180900_add_dry_to_user_migration_import.rb]: I voted for `detault: true` when I thought `dry` only affected the database check. As this finally also affects dbm/redis import, I agree that dry should default to false.

File: app/models/carto/user_migration_import.rb (hunk 14 above)
[the remark this answers is not in the document]
Comment (javitonino) [reply app/models/carto/user_migration_import.rb]: `import_metadata` here should just be unconditionally `false`. This is the part of _user mover_ that changes the database host. I think it should always be disabled, so UserMigrationImport has the full responsibility of that.

Note that this has nothing to do with the metadata importing (database models). Naming is just very confusing.

[the remark this answers is not in the document]
Comment (juanignaciosl) [reply app/models/carto/user_migration_import.rb]: > The default for import_metadata is true at the (user mover)

There's no `import_metadata` parameter there, so this might be a bug. Maybe you wanted `metadata`? Please confirm this, because if I'm right, this is always importing the metadata (test?).

That said, ok 👍 . I didn't realize that this was related to Redis and that 👍

File: services/user-mover/utils.rb (hunk 12 above)
Comment (juanignaciosl) [reply services/user-mover/utils.rb]: Ok, that's a legit reason. Maybe we could output at the console the path to the file saying that "Check the progress at #{path}. The file will be dumped into DB log and deleted when process finishes".

Comment (juanignaciosl) [reply services/user-mover/utils.rb]: PD: but we should still take care with best practices.

Comment (amartinj) [reply services/user-mover/utils.rb]: @juanignaciosl I suggested that, but @javitonino was right when he said there should be no more than one migration -> one file in /tmp for the migration

Comment (juanignaciosl) [reply services/user-mover/utils.rb]: Yep, but it has no cost, it's explicit, and documents the expected behaviour for my future me. _Pretty please_ 🤗

Comment (amartinj) [reply services/user-mover/utils.rb]: Block syntax example at the docs shows that closing explicitly is not needed in this case

**Comment by amartinj** [comment]
A new database host has been added to staging so we can test this. see [this issue](https://github.com/CartoDB/tech-ops/issues/306) for references.

File: app/models/carto/user_migration_import.rb [hunk 15]
```diff
@@ -26,8 +26,10 @@ class UserMigrationImport < ::ActiveRecord::Base
     validates :exported_file, presence: true
     validates :json_file, presence: true
     validate :valid_org_import
+    validate :valid_dry_settings
 
     def run_import
+      raise 'Incompatible options: import_metadata and dry' if import_metadata && dry
```
Comment (juanignaciosl) [inline app/models/carto/user_migration_import.rb]: @amartinj this is duplicated at the validation. If you want to explicitly fail on this, do something like `raise errors.full_messages.join(', ') unless valid?`

File: app/models/carto/user_migration_import.rb [hunk 16]
```diff
@@ -167,14 +173,14 @@ def import_job_arguments(data_dir)
         into_org_name: nil,
         mode: :import,
         logger: log.logger,
-        import_job_logger: log.logger
+        import_job_logger: log.logger,
+        update_metadata: !dry
```
Comment (javitonino) [inline app/models/carto/user_migration_import.rb]: I thought we agreed to set this always to `false` 😢. Let's do it another time, in order to avoid delays.

**Comment by javitonino** [comment]
Seems to work fine. I run into https://github.com/CartoDB/cartodb-central/issues/1945 during acceptance. Otherwise, it seems to work ok.
